### PR TITLE
`Process`: Allow `None` for input ports that are not required

### LINKED ISSUE #5720: `InputPort`: allow explicitly passed `None` for ports that are not required
Currently, when passing `None` to an input of a `Process` will raise a validation exception, even if that input is not required. This can be kind of surprising for a user. Consider the following:
```python
def accepts_none(a=None):
    pass

accepts_none()
accepts_none(None)
accepts_none(a=None)
```
All of these invocations work just fine, as one would expect. The same holds if `accepts_none` is decorated with `calcfunction`. Process functions will also allow `None` for functions that define ``None`` as the default. The following, however, will raise:
```
class AcceptNoneProcess(Process):

    """Simple process with dynamic input namespace."""

    _node_class = orm.WorkflowNode

    @classmethod
    def define(cls, spec):
        super().define(spec)
        spec.input('not_required', valid_type=orm.Int, required=False)

run(AcceptNoneProcess)  # This works just fine
run(AcceptNoneProcess, not_required=None)  # But this will except
```
This is not consistent. We should consider making `None` an automatically accepted type for non-required input ports.

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #5720 

Python functions and process functions accept `None` for arguments
that define that as a default without any issues. That is to say,
the following works just fine:

    from aiida.engine import calcfunction

    @calcfunction
    def accepts_none(a=None):
        pass

    accepts_none()
    accepts_none(None)

For a `Process`, however, `None` is not accepted even if the port is not
required. This can be annoying for users because when writing wrappers
around a process launcher, they will have to manually check for one of
the inputs being `None` and remove it before submitting the input
dictionary.

Here we update the `InputPort` constructor to automatically add the
`NoneType` to the `valid_type` argument if the port is marked as
`required=False`. This way, non-required ports will accept `None`
without raising an exception during input validation.

### CODE REVIEW

File: tests/engine/test_process.py [hunk 1]
```diff
@@ -469,3 +469,25 @@ def test_input_validation_storable_nodes():
...
+
+
+@pytest.mark.usefixtures('clear_database_before_test')
+def test_not_required_accepts_none():
+    """Test that a port that is not required, accepts ``None``."""
+    from aiida.engine.utils import instantiate_process
+    from aiida.manage import get_manager
+
+    runner = get_manager().get_runner()
+    process = instantiate_process(runner, TestNotRequiredNoneProcess, not_required=None)
```
Comment (unkcpz) [inline tests/engine/test_process.py:492]: I think the process function with input parameter defaulting to `None` is not covered by this test, correct? Can you add to nightly tests which are exactly the process function manifest the case?

**Review by unkcpz — CHANGES_REQUESTED** [review]
@sphuber thanks! 
Maybe better to add a nightly daemon test for the process function with the parameter default set to `None` as what you describe in the commit message?

File: tests/engine/test_process.py (hunk 1 above)
Comment (sphuber) [reply tests/engine/test_process.py:492]: That is already being tested: https://github.com/aiidateam/aiida-core/blob/9c54b73da1664eab0279e323d3372fcf6102dcae/tests/engine/test_process_function.py#L244
Here we are adding the same behavior for any other process, such as `CalcJob` and `WorkChain`.

Comment (unkcpz) [reply tests/engine/test_process.py:492]: Ah, I see. The `FunctionProcess` is subclass of `Process` I thought this PR is going to fix the bug introduced by the previously added feature for `None` supported input. 
It is all good to me! Thanks!

**Review by unkcpz — APPROVED** [review]

**Comment by mbercx** [comment]
@sphuber unfortunately I think this change has had some unintended backwards-incompatible side effects. When trying to run a `PwBaseWorkChain` with `aiida-core==2.1.0`, the process excepts with the following traceback:

```python
$ verdi process report 12099
2022-11-11 16:33:57 [4372 | REPORT]: [12099|PwBaseWorkChain|on_except]: Traceback (most recent call last):
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/process_states.py", line 228, in execute
    result = self.run_fn(*self.args, **self.kwargs)
  File "/home/mbercx/envs/aiida-dev/code/aiida-core/aiida/engine/processes/workchains/workchain.py", line 252, in _do_step
    finished, stepper_result = self._stepper.step()
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/workchains.py", line 295, in step
    finished, result = self._child_stepper.step()
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/workchains.py", line 528, in step
    finished, result = self._child_stepper.step()
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/workchains.py", line 295, in step
    finished, result = self._child_stepper.step()
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/workchains.py", line 246, in step
    return True, self._fn(self._workchain)
  File "/home/mbercx/envs/aiida-dev/code/aiida-core/aiida/engine/processes/workchains/restart.py", line 197, in run_process
    node = self.submit(self.process_class, **inputs)
  File "/home/mbercx/envs/aiida-dev/code/aiida-core/aiida/engine/processes/process.py", line 523, in submit
    return self.runner.submit(process, *args, **kwargs)
  File "/home/mbercx/envs/aiida-dev/code/aiida-core/aiida/engine/runners.py", line 183, in submit
    process_inited = self.instantiate_process(process, *args, **inputs)
  File "/home/mbercx/envs/aiida-dev/code/aiida-core/aiida/engine/runners.py", line 169, in instantiate_process
    return instantiate_process(self, process, *args, **inputs)
  File "/home/mbercx/envs/aiida-dev/code/aiida-core/aiida/engine/utils.py", line 64, in instantiate_process
    process = process_class(runner=runner, inputs=inputs)
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/base/state_machine.py", line 194, in __call__
    inst.transition_to(inst.create_initial_state())
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/base/state_machine.py", line 336, in transition_to
    self.transition_failed(initial_state_label, label, *sys.exc_info()[1:])
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/base/state_machine.py", line 352, in transition_failed
    raise exception.with_traceback(trace)
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/base/state_machine.py", line 321, in transition_to
    self._enter_next_state(new_state)
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/base/state_machine.py", line 383, in _enter_next_state
    self._fire_state_event(StateEventHook.ENTERING_STATE, next_state)
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/base/state_machine.py", line 300, in _fire_state_event
    callback(self, hook, state)
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/processes.py", line 329, in <lambda>
    lambda _s, _h, state: self.on_entering(cast(process_states.State, state)),
  File "/home/mbercx/envs/aiida-dev/code/aiida-core/aiida/engine/processes/process.py", line 405, in on_entering
    super().on_entering(state)
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/processes.py", line 675, in on_entering
    call_with_super_check(self.on_create)
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/base/utils.py", line 29, in call_with_super_check
    wrapped(*args, **kwargs)
  File "/home/mbercx/envs/aiida-dev/code/aiida-core/aiida/engine/processes/process.py", line 395, in on_create
    super().on_create()
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/base/utils.py", line 16, in wrapper
    wrapped(self, *args, **kwargs)
  File "/home/mbercx/.virtualenvs/aiida-dev/lib/python3.8/site-packages/plumpy/processes.py", line 732, in on_create
    raise ValueError(result)
ValueError: Error occurred validating port 'inputs.settings': value 'settings' is not of the right type. Got '<class 'aiida.common.extendeddicts.AttributeDict'>', expected '(<class 'aiida.orm.nodes.data.dict.Dict'>, <class 'NoneType'>)'

2022-11-11 16:33:57 [4373 | REPORT]: [12099|PwBaseWorkChain|on_terminated]: remote folders will not be cleaned
```

The problem is that the `PwBaseWorkChain` uses the `prepare_process_inputs` function to convert e.g. the `settings` input into a `Dict` node, which in turn checks is `valid_type == Dict`. Since the `valid_type` is now converted in a list on the fly, this breaks the `PwBaseWorkChain`.

I was looking into fixing this in the `aiida-quantumespresso` plugin, but perhaps it is better to revert this commit? Other plugins might also rely on checking the `valid_type` of an input port during runtime, and hence break as well.

**Comment by sphuber** [comment]
Thanks for letting us know @mbercx . 

I think this was actually really an inaccuracy that was anyway hiding in the `BaseRestartWorkChain`. So why don't we just fix that by changing
```python
elif port.valid_type == orm.Dict and isinstance(value, dict):
```
to
```python
valid_types = port.valid_type if isinstance(port.valid_type, (list, tuple)) else (port.valid_type,)

...
elif orm.Dict in valid_types and isinstance(value, dict):
```
Do you really think there is other code out there that might break due to this?

**Comment by mbercx** [comment]
Well, the QE plugin also does ^^

https://github.com/aiidateam/aiida-quantumespresso/blob/d15f8cd1032e37ae59a19d733ee5ae1ed59e1f10/src/aiida_quantumespresso/utils/mapping.py#L94-L95

As I said in https://github.com/aiidateam/aiida-core/pull/5757, I'm not sure if there is a lot of other plugins that rely on checking the `valid_type` on the fly. I suppose it's a rather rare use case.
